org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -283,16 +285,27 @@ private void populateDataFields(List<Either<Command, CodeAction>> codeActions) {
 	}
 
 	public static IProblemLocationCore[] getProblemLocationCores(ICompilationUnit unit, List<Diagnostic> diagnostics) {
-		IProblemLocationCore[] locations = new IProblemLocationCore[diagnostics.size()];
+		List<IProblemLocationCore> locations = new ArrayList<>();
 		for (int i = 0; i < diagnostics.size(); i++) {
 			Diagnostic diagnostic = diagnostics.get(i);
 			int problemId = getProblemId(diagnostic);
-			int start = DiagnosticsHelper.getStartOffset(unit, diagnostic.getRange());
-			int end = DiagnosticsHelper.getEndOffset(unit, diagnostic.getRange());
-			boolean isError = diagnostic.getSeverity() == DiagnosticSeverity.Error;
-			locations[i] = new ProblemLocationCore(start, end - start, problemId, new String[0], isError, IJavaModelMarker.JAVA_MODEL_PROBLEM_MARKER);
+			boolean handled = false;
+			if (problemId == IProblem.UndefinedName) {
+				for (Diagnostic d : diagnostics) {
```

Comment:
Instead of having an additional for-loop here, to check if we're comparing 2 diagnostics we know will resolve the issue in the same way, and on the same line, could we move this code into [`handledProblems`](https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corrections/QuickFixProcessor.java#L73) ? Would be good to have that kind of logic all in one spot.

I guess you could override ArrayList so that add takes into account our idea of "identical" problems. Another alternative is to just create a helper method that uses the code from here just before calling List.add(..). Only thing missing is the problem arguments need to be passed into the IProblemLocationCore.